src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -2341,7 +2341,7 @@ final boolean needsConversionToLambda() {
                         receiverIsReferenceProjection() ||
                         (tree.getMode() == ReferenceMode.NEW &&
                           tree.kind != ReferenceKind.ARRAY_CTOR &&
-                          (tree.sym.owner.isDirectlyOrIndirectlyLocal() || tree.sym.owner.isInner() || tree.sym.owner.isPrimitiveClass()));
+                          (tree.sym.owner.isDirectlyOrIndirectlyLocal() || tree.sym.owner.isInner() || tree.sym.owner.isValueClass()));
             }
```

Comment:
Notes for reviewer: Basically by forcing the constructor references to be converted into lambdas, we ensure that later pipeline stages will see the constructor invocation. For example, with the code change above, given a value class V, a constructor reference of the form V::new will be transformed into a lambda implementation method of the form 

```
/*synthetic*/ private static X$V lambda$main$1() {
    return new X$V();
}
```
This lambda implementation method has the constructor invocation intact which can then be detected and transformed by *com.sun.tools.javac.jvm.TransValues#visitNewClass*

(It is possible to directly emit BSM entries without converting them to lambda but that would result in code duplication - we would have to look at DynamicMethod symbols, pick them apart and reencode static and dynamic arguments to indy so that value constructors (and primitive class constructors are reencoded to static factory)

Reply:
I wonder if the transformation from a constructor call to a static method call when it's a value class should not be done before the lambda transformation ?

Reply:
> I wonder if the transformation from a constructor call to a static method call when it's a value class should not be done before the lambda transformation ?

This is technically feasible, but has some practical complications. As the top level comment in com.sun.tools.javac.jvm.TransValues calls out:

```
/**
 * This pass translates value class constructors into static factory methods and patches up constructor
 * calls to become invocations of those static factory methods.
 *
 * We get commissioned as a subpass of Gen. Constructor trees undergo plenty of change in Lower
 * (enclosing instance injection, captured locals ...) and in Gen (instance field initialization,
 * see normalizeDefs) and so it is most effective to wait until things reach a quiescent state
 * before undertaking the tinkering that we do.
 *
 * See https://bugs.openjdk.java.net/browse/JDK-8198749 for the kind of transformations we do.
 *
 */
```

Lambda lowering happens well before Lower and Gen pipeline phases. TransValues is a subphase of Gen - we need to wait until Gen because constructors undergo lots of changes up until and well into Gen. So if we want to group all transformations - of constructors into static factories and constructor invocations into static value factory invocations - into one centralized localized pipeline stage (i.e TransValues) the present approach is the best

Reply:
Okay, make sense